lib/bots/keyframebot.ts

```diff
@@ -0,0 +1,684 @@
+    private environments: Environments;
+    /** Product repo. */
+    private productRepo: string;
+
+    /**
+     * Constructor for the KeyframeBot.
+     * @param name          Name of the KeyframeBot.
+     * @param constObject   Constructor object.
+     */
+    constructor(name: string, constObject: KeyframeBotConstructor) {
+        // This is the KeyframeBot.
+        super(name);
+
+        const integrationId = constObject.integrationId;
+        const pemString = constObject.pem;
+        const webhook = constObject.webhookSecret;
+        const environments = constObject.environments;
+
+        // Create a new listener for Github with the right Integration ID.
+        const ghListener = this.addServiceListener('github', {
+            client: name,
+            loginType: {
+                integrationId,
+                pem: pemString,
+                type: 'integration'
+            },
+            path: '/keyframehooks',
+            port: GithubPort,
+            type: 'listener',
+            webhookSecret: webhook
+        });
+
+        // Create a new emitter with the right Integration ID.
+        const ghEmitter = this.addServiceEmitter('github', {
+            loginType: {
+                integrationId,
+                pem: pemString,
+                type: 'integration'
+            },
+            pem: pemString,
+            type: 'emitter'
+        });
+
+        // Throw if we didn't get either of the services.
+        if (!ghListener) {
+            throw new Error("Couldn't create a Github listener");
+        }
+        if (!ghEmitter) {
+            throw new Error("Couldn't create a Github emitter");
+        }
+        this.githubListenerName = ghListener.serviceName;
+        this.githubEmitterName = ghEmitter.serviceName;
+
+        // Github API handle
+        this.githubApi = (<GithubHandle>ghEmitter.apiHandle).github;
+        if (!this.githubApi) {
+            throw new Error('No Github API instance found');
+        }
+
+        // Parse our environments.
+        try {
+            this.environments = JSON.parse(constObject.environments);
+        } catch (err) {
+            throw new Error('There are no valid environments to use');
+        }
+        this.logger.log(LogLevel.INFO, `---> ${name}: Aware of the following environments: ${environments}`);
+
+        // Get the product repo.
+        this.productRepo = constObject.productRepo;
+
+        // Create a new endpoint to allow keyframes to be promoted to a particular environment.
+        // New Express app. We'll reuse it in the GH SL.
+        this.expressApp = express();
+        if (!this.expressApp) {
+            throw new Error("Couldn't create an Express application");
+        }
+
+        // Add body parser.
+        this.expressApp.use(bodyParser.urlencoded({ extended: true }));
+        this.expressApp.use(bodyParser.json());
+        this.expressApp.post(DeployKeyframePath, this.deployKeyframe);
```

Comment:
This falls under deployment. Almost all of our services listen to HTTP and then get HTTPS routes setup on top of them by a front-facing system (such as nginx).